### PR TITLE
deconv op implementing ...

### PR DESCRIPTION
**Pull request by zchen0211**

### CODE REVIEW

File: paddle/operators/deconv2d_op.h [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    int M = input->dims()[1];
+    int H = input->dims()[2];
+    int W = input->dims()[3];
+
+    int K_H = filter.dims()[2];
+    int K_W = filter.dims()[3];
+
+    int C = output_grad->dims()[1];  // output channels
+    int O_H = output_grad->dims()[2];
+    int O_W = output_grad->dims()[3];
```
Comment (chengduoZH) [inline paddle/operators/deconv2d_op.h]: [variable names style](https://google.github.io/styleguide/cppguide.html#Variable_Names)

File: paddle/operators/deconv2d_op.cc [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                                 framework::OpAttrChecker* op_checker)
+    : OpProtoAndCheckerMaker(proto, op_checker) {
+  AddInput(
+      "Input",
+      "The input tensor of deconvolution operator. "
+      "The format of input tensor is NMHW. Where N is batch size, M is the "
+      "number of input channels, H and W is the height and width of image.");
+  AddInput("Filter",
+           "The filter tensor of deconvolution operator."
+           "The format of the filter tensor is MCHW, where M is the number of "
```
Comment (chengduoZH) [inline paddle/operators/deconv2d_op.cc]: "MCHW" - >"MCHW"

File: paddle/operators/deconv2d_op.cc [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+           "input image channels, C is the number of output image channels, "
+           "H and W is height and width of filter. "
+           "We enforce groups number == 1 and padding == 0 in our "
+           "deconvolution Scenario.");
```
Comment (chengduoZH) [inline paddle/operators/deconv2d_op.cc]: We enforce groups number == 1 and padding == 0 in ~~our~~ deconvolution Scenario.

File: paddle/operators/deconv2d_op.cc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  std::vector<int> strides = ctx->Attrs().Get<std::vector<int>>("strides");
+  std::vector<int> paddings = ctx->Attrs().Get<std::vector<int>>("paddings");
+
+  for (int i = 0; i < paddings.size(); ++i) {
+    PADDLE_ENFORCE_EQ(paddings[i], 0, "No Padding allowed in deconv op.");
+  }
+
+  PADDLE_ENFORCE_EQ(in_dims.size(), 4, "Deconv2DOp input should be 4-D.");
+  PADDLE_ENFORCE_EQ(filter_dims.size(), 4, "Deconv2DOp filter should be 4-D.");
+  PADDLE_ENFORCE_EQ(in_dims[1], filter_dims[0],
```
[the remark this answers is not in the document]
Comment (zchen0211) [reply paddle/operators/deconv2d_op.cc]: Done.

File: paddle/operators/deconv2d_op.cc [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+Deconv2DOpMaker::Deconv2DOpMaker(framework::OpProto* proto,
+                                 framework::OpAttrChecker* op_checker)
+    : OpProtoAndCheckerMaker(proto, op_checker) {
+  AddInput(
+      "Input",
+      "The input tensor of deconvolution operator. "
+      "The format of input tensor is NMHW. Where N is batch size, M is the "
+      "number of input channels, H and W is the height and width of image.");
+  AddInput("Filter",
```
[the remark this answers is not in the document]
Comment (zchen0211) [reply paddle/operators/deconv2d_op.cc]: Done.

File: paddle/operators/deconv2d_op.cc (hunk 2 above)
Comment (zchen0211) [reply paddle/operators/deconv2d_op.cc]: Done.

File: paddle/operators/deconv2d_op.cc (hunk 3 above)
Comment (zchen0211) [reply paddle/operators/deconv2d_op.cc]: Done.

File: paddle/operators/deconv2d_op.h [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+ protected:
+  void InferShape(framework::InferShapeContext* ctx) const override;
+};
+
+template <typename Place, typename T>
+class GemmDeconv2DKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    const Tensor* input = context.Input<Tensor>("Input");
+    // filter will be reshaped, so we do not use constant pointer here
```
[the remark this answers is not in the document]
Comment (zchen0211) [reply paddle/operators/deconv2d_op.h]: Done.

File: paddle/operators/deconv2d_op.h [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+template <typename Place, typename T>
+class GemmDeconvGrad2DKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    const Tensor* input = context.Input<Tensor>("Input");
+    const Tensor* output_grad =
+        context.Input<Tensor>(framework::GradVarName("Output"));
+
+    // For filter, we do not use const pointer b/c we will do reshape
+    // but we should avoid modifying its value
```
[the remark this answers is not in the document]
Comment (zchen0211) [reply paddle/operators/deconv2d_op.h]: Done.

File: paddle/operators/deconv2d_op.h [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    Tensor filter = *context.Input<Tensor>("Filter");
+
+    Tensor* input_grad =
+        context.Output<Tensor>(framework::GradVarName("Input"));
+    Tensor* filter_grad =
+        context.Output<Tensor>(framework::GradVarName("Filter"));
+
+    std::vector<int> strides = context.Attr<std::vector<int>>("strides");
+    // Actually, no paddings and groups allowed in deconv
```
[the remark this answers is not in the document]
Comment (zchen0211) [reply paddle/operators/deconv2d_op.h]: Done.

File: paddle/operators/deconv2d_op.cc [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#include "paddle/operators/deconv2d_op.h"
+#include "paddle/operators/conv2d_op.h"
+
+namespace paddle {
+namespace operators {
+
+void Deconv2DOp::InferShape(framework::InferShapeContext* ctx) const {
```
Comment (qingqing01) [inline paddle/operators/deconv2d_op.cc]: From the comments: https://github.com/tensorflow/tensorflow/issues/256#issuecomment-162257789

How about rename Conv2DTranspose?

File: paddle/operators/deconv2d_op.cc [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+Deconv2DOpMaker::Deconv2DOpMaker(framework::OpProto* proto,
+                                 framework::OpAttrChecker* op_checker)
+    : OpProtoAndCheckerMaker(proto, op_checker) {
+  AddInput(
+      "Input",
+      "The input tensor of deconvolution operator. "
+      "The format of input tensor is NMHW. Where N is batch size, M is the "
+      "number of input channels, H and W is the height and width of image.");
```
Comment (qingqing01) [inline paddle/operators/deconv2d_op.cc]: "(Tensor) The input tensor of transposed 2D convolution operator. "

The () is used to denote the type, same as the following annotations.
 
NMHW -> NCHW

File: paddle/operators/deconv2d_op.h [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include "glog/logging.h"
```
Comment (qingqing01) [inline paddle/operators/deconv2d_op.h]: remove glog.

File: paddle/operators/deconv2d_op.h [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+
+    // deconvolution: gemm + col2im (similar to conv-backward on input)
+
+    output->mutable_data<T>(context.GetPlace());
+    auto t = framework::EigenVector<T>::Flatten(*output);
+    t.device(context.GetEigenDevice<Place>()) = t.constant(static_cast<T>(0));
+
+    for (int i = 0; i < N; i++) {
+      // batch with size (M, H * W)
+      Tensor input_batch = input->Slice<T>(i, i + 1).Resize(input_matrix_shape);
```
Comment (qingqing01) [inline paddle/operators/deconv2d_op.h]: Update code, since the `Slice` removed the template T.

File: paddle/operators/deconv2d_op.h [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+  void Compute(const framework::ExecutionContext& context) const override {
+    const Tensor* input = context.Input<Tensor>("Input");
+    // filter will be reshaped, so we do not use constant pointer here
+    Tensor filter = *context.Input<Tensor>("Filter");
+
+    Tensor* output = context.Output<Tensor>("Output");
+
+    std::vector<int> strides = context.Attr<std::vector<int>>("strides");
+
+    // no paddings and groups allowed in deconv
```
Comment (qingqing01) [inline paddle/operators/deconv2d_op.h]: If to do in next PR, add TODO comments.

File: paddle/operators/deconv2d_op.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (zchen0211) [reply paddle/operators/deconv2d_op.cc]: The Deconv case is a little different from Conv case. Like in Caffe2, Conv2d use NCHW for input and MCHW for filter and produces a tensor of shape NMHW; Caffe2 Deconv applies NCHW for input, CMHW for filter and produces output tensor with shape NMHW. I will make it clear in my codes.

File: paddle/operators/deconv2d_op.cc [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+                 "Output(Output) of Deconv2DOp should not be null.");
+
+  auto in_dims = ctx->GetInputDim("Input");
+  auto filter_dims = ctx->GetInputDim("Filter");
+  std::vector<int> strides = ctx->Attrs().Get<std::vector<int>>("strides");
+  std::vector<int> paddings = ctx->Attrs().Get<std::vector<int>>("paddings");
+
+  for (int i = 0; i < paddings.size(); ++i) {
+    PADDLE_ENFORCE_EQ(paddings[i], 0, "No Padding allowed in deconv op.");
+  }
```
[the remark this answers is not in the document]
Comment (Canpio) [reply paddle/operators/deconv2d_op.cc]: For `vector` is not supported by `AttrChecker` right now, we can only do check jobs in the infer shape.

File: paddle/operators/deconv2d_op.h [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+    // use col_matrix_shape in the gemm calculation
+    DDim col_matrix_shape = {C * K_H * K_W, H * W};
+
+    Tensor col;
+    col.mutable_data<T>(col_shape, context.GetPlace());
+    // col_matrix shares the same piece of data with col,
+    // but will be reshaped into a two-dimensional matrix shape
+    // to call the matrix multiplication interface.
+    Tensor col_matrix = col;
+    col_matrix.Resize(col_matrix_shape);
```
Comment (mkliegl) [inline paddle/operators/deconv2d_op.h]: That copy assign works as intended, but i looks a little unnatural to me at first glance, since for e.g. std::vector<T>, copy assign copies the data. However, copy assignment does share data in this case because the data is stored inside a  std::shared_ptr<Placeholder> inside the Tensor class. Nevertheless, I would suggest the more explicit:
```
Tensor col_matrix;
col_matrix.ShareDataWith(col);
```
(I realize this is carried over from conv2d_op.h - maybe you could change it there, too?)

File: paddle/operators/deconv2d_op.h [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+
+    DDim filter_matrix_shape = {M, C * K_H * K_W};
+    filter.Resize(filter_matrix_shape);
+
+    // deconvolution grad on input:
+    // im2col + gemm (similar to conv-forward)
+    // input need to compute gradient
+    if (input_grad) {
+      Tensor col_matrix = col;
+      DDim col_matrix_shape = {C * K_H * K_W, H * W};
```
Comment (mkliegl) [inline paddle/operators/deconv2d_op.h]: See above comment. I would prefer the more explicit:
```
Tensor col_matrix;
col_matrix.ShareDataWith(col);
```

File: paddle/operators/deconv2d_op.cc [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+           "input image channels, C is the number of output image channels, "
+           "H and W is height and width of filter. "
+           "We enforce groups number == 1 and padding == 0 in our "
+           "deconvolution Scenario.");
+  AddOutput("Output",
+            "The output tensor of deconvolution operator."
+            "The format of output tensor is also NCHW.");
+  AddAttr<std::vector<int>>("strides", "strides of deconvolution operator.")
+      .SetDefault({1, 1});
+  AddAttr<std::vector<int>>("paddings", "paddings of deconvolution operator.")
```
[the remark this answers is not in the document]
Comment (zchen0211) [reply paddle/operators/deconv2d_op.cc]: As @Canpio said, for current version to pass, we temporarily put our check here.

File: paddle/operators/deconv2d_op.h [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+    int M = input->dims()[1];
+    int H = input->dims()[2];
+    int W = input->dims()[3];
+
+    int K_H = filter.dims()[2];
+    int K_W = filter.dims()[3];
+
+    int C = output->dims()[1];  // output channels
+    int O_H = output->dims()[2];
+    int O_W = output->dims()[3];
```
[the remark this answers is not in the document]
Comment (zchen0211) [reply paddle/operators/deconv2d_op.h]: Done.

File: paddle/operators/deconv2d_op.h (hunk 1 above)
Comment (zchen0211) [reply paddle/operators/deconv2d_op.h]: Done.

File: paddle/operators/deconv2d_op.cc (hunk 10 above)
Comment (zchen0211) [reply paddle/operators/deconv2d_op.cc]: Done.

File: paddle/operators/deconv2d_op.cc (hunk 9 above)
Comment (zchen0211) [reply paddle/operators/deconv2d_op.cc]: Great suggestion!

File: paddle/operators/deconv2d_op.h (hunk 12 above)
Comment (zchen0211) [reply paddle/operators/deconv2d_op.h]: Done.

File: paddle/operators/deconv2d_op.h (hunk 15 above)
Comment (zchen0211) [reply paddle/operators/deconv2d_op.h]: Great thanks!

File: paddle/operators/deconv2d_op.h (hunk 11 above)
Comment (zchen0211) [reply paddle/operators/deconv2d_op.h]: Done.

File: paddle/operators/deconv2d_op.h (hunk 13 above)
Comment (zchen0211) [reply paddle/operators/deconv2d_op.h]: Done.

File: paddle/operators/deconv2d_op.h (hunk 15 above)
Comment (zchen0211) [reply paddle/operators/deconv2d_op.h]: Done.

File: paddle/operators/deconv2d_op.h (hunk 16 above)
Comment (zchen0211) [reply paddle/operators/deconv2d_op.h]: Done.

File: paddle/operators/conv2dtranspose_op.cc [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+                            "strides of convolution transpose operator.")
+      .SetDefault({1, 1});
+  AddAttr<std::vector<int>>("paddings",
+                            "paddings of convolution transpose operator.")
+      .SetDefault({0, 0});
+  AddComment(R"DOC(
+The convolution transpose operation calculates the output based on the input, filter
+and strides, paddings, groups parameters. The size of each dimension of the
+parameters is checked in the infer-shape.
+)DOC");
```
Comment (chengduoZH) [inline paddle/operators/conv2dtranspose_op.cc]: Better to give how to calculate the output height/width according to the input height/with, padding and stride size.

The doc of Pytorch is good.

http://pytorch.org/docs/master/nn.html#convtranspose2d

File: paddle/operators/conv2dtranspose_op.h [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+    const Tensor* input = context.Input<Tensor>("Input");
+    // The filter will be reshaped, so it should not be constant pointer
+    Tensor filter = *context.Input<Tensor>("Filter");
+
+    Tensor* output = context.Output<Tensor>("Output");
+
+    std::vector<int> strides = context.Attr<std::vector<int>>("strides");
+
+    // TODO(Zhuoyuan): Paddings can be added in future.
+    // groups will alway be disabled in conv2dtranspose.
```
Comment (chengduoZH) [inline paddle/operators/conv2dtranspose_op.h]: The attribution of group should be available in conv2dtranspose. Reference [Pytorch](http://pytorch.org/docs/master/nn.html#convtranspose2d)

File: paddle/operators/conv2dtranspose_op.h [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+    // convolution transpose: gemm + col2im (similar to conv-backward on input)
+
+    output->mutable_data<T>(context.GetPlace());
+    auto t = framework::EigenVector<T>::Flatten(*output);
+    t.device(context.GetEigenDevice<Place>()) = t.constant(static_cast<T>(0));
+
+    for (int i = 0; i < batch_size; i++) {
+      // batch with size (M, h * w)
+      Tensor input_batch = input->Slice(i, i + 1).Resize(input_matrix_shape);
+      // filter size: (M, c * k_h * k_w)
```
Comment (chengduoZH) [inline paddle/operators/conv2dtranspose_op.h]: You can delete this comment or write in line 119.

File: paddle/operators/conv2dtranspose_op.h [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+
+      input_grad->mutable_data<T>(context.GetPlace());
+      auto t = framework::EigenVector<T>::Flatten(*input_grad);
+      t.device(context.GetEigenDevice<Place>()) = t.constant(static_cast<T>(0));
+
+      for (int i = 0; i < batch_size; i++) {
+        // batch with size (c, o_h * o_w)
+        Tensor output_grad_batch =
+            output_grad->Slice(i, i + 1).Resize(output_shape);
+        // filter of size (m, c * k_h * k_w)
```
Comment (chengduoZH) [inline paddle/operators/conv2dtranspose_op.h]: Same as above.

File: python/paddle/v2/framework/tests/test_conv2dtranspose_op.py [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+            max_relative_error=0.05,
+            no_grad_set=set(['Input']))
+
+    def init_test_case(self):
+        self.pad = [0, 0]
+        self.stride = [1, 1]
+        self.dilations = [1, 1]
+        self.input_size = [2, 3, 5, 5]  # NCHW
+        f_c = self.input_size[1]
+        self.filter_size = [f_c, 6, 3, 3]
```
Comment (chengduoZH) [inline python/paddle/v2/framework/tests/test_conv2dtranspose_op.py]: If the calculation is not very time-consuming, you can write several border test examples.
